---
title: "Microbiome_analysis"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document that describes what the functions in cedrics microbiome pipeline are doing. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

# [**util.R**]{style="color:green"} - Loading the microbiome data into phyloseq

Cedric developed a simple script for loading the microbiome data into phyloseq, it includes path to all of the necessary files (Need to download first the labhuiofrank helper scripts in: <https://github.com/labhuiofrank/16S-data-analysis.git>)

The source command is going to run another r script called loading.R, on your lab mac site this file is located at "bootcamp-16S/labhuiofrank-16S-util/R/loading.R" The loading.R file is a set of functions (load_shared, load_counttable, load_constaxonmy) that you will need to format the data into a phyloseq object.

**If you change the name or location of your data or your metadata you will need to modify this code and make sure it goes to the appropriate file path**

### You do not need to run this util.R code it will run automatically with the next section

```{r, echo = TRUE}
source("/Users/klfrank/Dropbox/Post-Grad/Bioinformatics/R-Microbiome-Statistics/Databases/REU-bootcamp/labhuiofrank-16S-util/R/loading.R")
library(phyloseq)

# path to data
abundance_file <- "/Users/klfrank/Dropbox/Post-Grad/Protocols/Bootcamp/Bioinformatics/data/OTUs.100.rep.count_table"
taxonomy_file <- "/Users/klfrank/Dropbox/Post-Grad/Protocols/Bootcamp/Bioinformatics/data/OTUs.100.cons.taxonomy"
metadata_file <- "/Users/klfrank/Dropbox/Post-Grad/Protocols/Bootcamp/Bioinformatics/data/metadata.csv"
metadata <- read.table(metadata_file, sep=",", row.names=1, header=T) 
#if you need to reorder your variables so your graphs look nice, you would do this here
GROUP_ahupuaa <-c("Before_Kawainui", "Kawainui", "After_Kawainui", "Kailua_Bay")
metadata$Ahupuaa_Location <- factor(metadata$Ahupuaa_Location, levels=GROUP_ahupuaa)
#if you need to reorder your variables so your graphs look nice, you would do this here

# Loader function to import for each analyses
load_microbiome <- function() {
    abundance <- load_counttable(abundance_file)
    taxonomy <- load_constaxonomy(as.matrix(taxonomy_file))
    metadata <- metadata 
      
    ps <- phyloseq(
        otu_table(abundance, taxa_are_rows=T),
        tax_table(as.matrix(taxonomy)),
        sample_data(metadata)
    )
    return(ps)
}

#' Custom palette function for large categorical groups
#' Repeats the palette if more than 50 groups
get_palette <- function(n) {
  palette <- c("#45cc65","#8e3ebc","#62ba3f","#9164dc","#a75355","#91ba33","#cd74e5","#33a150","#b8319c","#4dcc90","#e75dc1","#538d2a","#586cd8","#dca831","#9049a2","#b2b33a","#d57ccb","#80bc6c","#e54486","#4fcfc3","#df4b40","#4db9df","#de632b","#5e91d3","#d68330","#4363a5","#9c8026","#ae99e4","#60751c","#7961a8","#3d864a","#af3a77","#66b78b","#c3364e","#2ba198","#aa4b28","#35896b","#e2798a","#1a6447","#e390bf","#306a3c","#a56395","#4d6d30","#924869","#aab56e","#e38f6b","#666020","#d1a965","#966433","#868949")
  if (n > 50) {
    palette <- rep(palette, 1+floor(n/50))[1:n]
  }
  return(palette)
}

options(ggplot2.discrete.colour=RColorBrewer::brewer.pal(9, "Set1"))
options(ggplot2.discrete.fill=RColorBrewer::brewer.pal(9, "Set1"))
```

# [**alpha_diversity.R**]{style="color:green"}- Alpha diversity

The following code is run in the alpha_diversity file. Iʻve chunked it up here so I can explain what is going on.

```{r, echo=TRUE}
#source("bootcamp-16S/util.R")
library("ggplot2")
library("speedyseq") # phyloseq extension

# Global variable
# What are we focusing on?
target <- "Ahupuaa_Location"
```

## =====================

## Load our data

## =====================

First you want to load all our data into a phyloseq object.

Cedric made a really easy function to do this with load_microbiome(), it formats all the data files and puts them here. To get a better feel for the data we have you can just type the varible "mb" and it will tell you how many taxa (a.k.a ASVs - Amplicon Sequence Variants) in the number of samples you have. It will also tell you how many columns (a.k.a sample variables) you have in your metadata file. In the case of our Kailua data mb has 33637 ASVs in 193 samples and there are 44 sample variables in our metadata file.

```{r, echo=TRUE}

mb <- load_microbiome()
mb

```

## =====================

## Preprocess data

## =====================

Before we do any further analysis we need to process the data. This will help us to merge or subset our data.

Merging data will pool your data set into a smaller number of samples - for example if you have 15 samples which are 3 replicates of 2500 sequences from 5 sites, if you merge by site you will now have 5 samples each with 7500 sequences.

However, if you just want to look at say the mud samples or the water samples you will want to subset your data. For eaxample if you have 15 samples, 5 of which are mud and 10 of which are water, you can subset your samples for just the mud and you will now have only 5 mud samples in your dataset.

```{r, echo=TRUE}
## 1. MERGE samples: if you want to merge samples this is how to do it, for example the following merges all the replicates into one sample. "Sample_no" is a column in the metadata.  In our Kailua data - if you want to merge both the 0.2 and 0.8um filters you can merge by "ID" 
mb_merge <- merge_samples2(mb, "Sample_no")
mb_merge

## 2. SUBSET (you can subset for anything in any column of the metadata)
mb_mud <- subset_samples(mb, Sample_Type %in% "Mud")
mb_mud
mb_water <- subset_samples(mb, Sample_Type %in% "Water")
mb_water
## you can even subset by taxa, for example you could subset all the human pathogens in the system
pathogens <- read.table("/Users/klfrank/Dropbox/Post-Grad/Bioinformatics/Projects/COBRE/R_16S_analysis/Input/pathogens_genus.csv", header=T, row.names=1, sep=",")
mb_path <- subset_taxa(mb, Genus %in% rownames(pathogens))
mb_path
```

[**Just by looking at the summary of mb_path you can see that there are 931 ASVs of putative human pathogens in the 193 samples from Kailua**]{style="color:green"}

## =================================

## Display in barplot with phyloseq

## =================================

helpful code for presentations or reports

## [**theme_presentation.R**]{style="color:green"}

```{r, echo=TRUE}
#code for presentations/reports
source("/Users/klfrank/Dropbox/Post-Grad/Bioinformatics/R-Microbiome-Statistics/Functions/Function_Scripts/theme_presentation.R")
```

This will prep and display data based on the target that you are interested, the code merges the sample based on the target of interest and then transforms the data to relative abundance. To simplify the bar graph we will only keep the ASVs that make up at least 1% of the total abundance. Then we plot the bar graph - in the fill position of plot_bar you can change to anything from Kindgom, Phylum, Class, Order, Family, Genus

```{r, echo=TRUE}
# Global variable
# What are we focusing on?
target <- "Ahupuaa_Location"

mb_per_type <- merge_samples2(mb_water, target)
mb_per_type <- transform_sample_counts(mb_per_type, function(x) x/sum(x))

# Keep ASVs that make at least 1% of the total abundance
mb_core <- filter_taxa(mb_per_type, function(x) any(x > 0.01), TRUE)
#mb_core <- transform_sample_counts(mb_core, function(x) x/sum(x)) #this sets everything to 100% after you remove ASVS

taxa_bargraph <- plot_bar(mb_core, target, fill="Class") +
  geom_bar(aes(fill=Class), stat="identity", position="stack") +
  scale_fill_manual(values=get_palette(ntaxa(mb_core)))
taxa_bargraph 
taxa_bargraph+theme_report(tsize=16)
```

Here is a different function to plot your taxa_barplots, instead of only keeping ASVs that are at least 1% of the total abundance, this one groups any ASVs that are less than 1% of the abundance as other \## \*\* [taxa_barplot.R]{style="color:green"}\*\*

```{r, echo=TRUE}
source("/Users/klfrank/Dropbox/Post-Grad/Bioinformatics/R-Microbiome-Statistics/Functions/Function_Scripts/taxa_barplot.R")

taxa_bp <- taxa_barplot(mb_water,"Class")
taxa_bp + theme_report()
```

Lets say you wanted to look at the abundance of just a subset, like human pathogens. This will work with any subset file, you just need a list of taxa at the genus level

## [**Consistent_color_pallete.R**]{style="color:green"}

## [**subset_barplot.R**]{style="color:green"}

```{r, echo=TRUE}
#set color paletes
source("/Users/klfrank/Dropbox/Post-Grad/Bioinformatics/R-Microbiome-Statistics/Functions/Function_Scripts/Consistent_Pallet/consistent_color_pallete.R")
#upload functions for subsetting barplot
source("/Users/klfrank/Dropbox/Post-Grad/Bioinformatics/R-Microbiome-Statistics/Functions/Function_Scripts/subset_barplot.R")
path.bar <- subset_barplot(mg=mb_water, tax_lvl="Genus", groups=target, thresh=1e-4, nrows_legend=40)
path.bar+ theme_report(tsize=16)+ggtitle("Relative Abundance of Putative Human Pathogens in Kailua")
```

## ========================

## Alpha diversity metrics

## ========================

Alpha diversity (α-diversity) is defined as the mean diversity of species in different sites or habitats within a local scale.

Species richness: Chao1 Index (chao) is an estimator based on abundance; thus, it requires data that refers to the abundance of individual samples belonging to a certain class. Higher chao \~ higher diversity.

Community evenness: Refers to how close in numbers each species in an environment is

Community Diversity indices provide more information about community composition than simply species richness as they also take the relative abundances of different species into account providing important information about rarity and commonness of species in a community. The Shannon Index (shannon) takes into account the number of species living in a habitat (richness) and their relative abundance (evenness).Lower shannon \~ higher diversity. An index value of 1 means that all groups have the same frequency.

InvSimpson parameter is preferred to other measures of alpha-diversity because it is an indication of the richness in a community with uniform evenness that would have the same level of diversity. So while measures such as the shannon index are somewhat abstract, the inverse of the Simpson index has some biological interpretation. Other advantages of the Simpson-based metrics are that they do not tend to be as affected by sampling effort as the shannon index.

```{r, echo=TRUE}
indices <- c("Chao1", "Shannon", "InvSimpson")

# One dot for each sample in group
plot_richness(mb, color=target, measures=indices)
# Or as a boxplot per site
plot_richness(mb, x=target, color=target, measures=indices) + 
  geom_boxplot()

# A bit prettier
plot_richness(mb, x=target, color=target, measures=indices) + 
  geom_boxplot(aes_string(fill=target), color="black") +
  scale_fill_brewer(palette="Set1") + scale_color_brewer(palette="Set1")
```

# [**beta_diversity.R**]{style="color:green"}

This is a seperate file that you can run to determine your beta_diversity; the front matter is very similar to the other file, and again you can subset/merge your samples in anyway you need

```{r, echo=TRUE}
#source("bootcamp-16S/util.R")
#library("ggplot2")
#library("speedyseq") # phyloseq extension
library("vegan")

# Global variable
# What are we focusing on?
target <- "Ahupuaa_Location"

## =====================##
##     Load our data    ##
## =====================##
mb <- load_microbiome()
```

## ======================

## Normalize samples

## ======================

Now we want to subsample for beta-diversity, this is REALLY REALLY important because we need all samples to have the same number of reads in order to analyze

Beta diversity quantifies the number of different communities in the region. So it's clear that beta diversity does not only account for the relationship between local and regional diversity, but also informs about the degree of differentiation among biological communities.

```{r, echo=TRUE}
# Compute sample counts
sample_counts <- sample_sums(mb_water)

# quickly
hist(log10(sample_counts), n=30)
threshold <- 4000

# Same, but a bit prettier
sample_sizes <- data.frame(x=sample_counts)

ggplot(sample_sizes, aes(x=x)) +
    geom_histogram(aes(y=..density..), alpha=0.5) + # bars
    geom_density(fill="blue", alpha=0.3) +          # curve
    geom_vline(xintercept=threshold, color="red", linetype=2) + # red dotted line to see the threshold
    scale_x_log10()

# Subsampling
mb_subsampled <- rarefy_even_depth(mb_water, threshold, rngseed=123, replace=FALSE)
mb_subsampled
```

## ==============================

## Unsupervised ordination: NMDS

## ==============================

This will squish your data into 2D. You will get the best results if you subset the data first into only what you are interested in looking at.

```{r, echo=TRUE}
nmds <- ordinate(mb_subsampled, method="NMDS", trymax=300, parallel=8)

plot_ordination(mb_subsampled, nmds, color=target) + 
  geom_point(size=3) +
  stat_ellipse(aes_string(x="NMDS1", y="NMDS2", fill=target), 
               alpha=0.1, show.legend=FALSE, level=0.5, geom="polygon") 
```

### Are the groupings above significant?

Here we run a PERMANOVA adonis test to see if the groupings above are significant. [**They are because p = 0.0001. The groupings describe 17.4% of the variability in the data (r2). Moreover each site is significantly different than eachother by pairwise PERMANOVA tests**]{style="color:green"}

```{r message=FALSE}
dists <- phyloseq::distance(mb_subsampled, method="bray")
meta <- data.frame(sample_data(mb_subsampled))

cat("
==================================
====== Global significance =======
==================================
")

adonis(dists ~ Ahupuaa_Location, data=meta, perm=9999)

cat("
===================================
====== Pairwise comparisons =======
===================================
")

pairwiseAdonis::pairwise.adonis2(dists ~ Ahupuaa_Location, data=meta, perm=9999)
```

## ============================

## Supervised ordination: CCA

## ============================

Canonical correlation analysis is used to identify and measure the associations among the microbial community structure, and other variables of interest like geochemistry. Its kinda like doing an ordnation for regressions.

```{r eval=FALSE, include=FALSE}
cca <- ordinate(mb_subsampled, method="CCA", formula=~ DO_percent + Temp_C + Sal_ppt + pH)

# A bit more complex: add the environmental variables as arrows
arrow_coords <- scores(cca, display = "bp") %>% data.frame %>% rownames_to_column("labels")
arrowhead = arrow(length = unit(0.05, "npc"))

plot_ordination(mb_subsampled, cca, type="Sample", color=target, shape="Ahupuaa_Location") + geom_point(size=3) +  
  geom_segment(aes(x=0, xend=CCA1, y=0, yend=CCA2, color=NULL, shape=NULL), arrow = arrowhead, data=arrow_coords) +
  geom_text(aes(x=1.2 * CCA1, y=1.2 * CCA2, label=labels, color=NULL, shape=NULL), data=arrow_coords)

# Other option: custom function from labhuiofrank github
source("/Users/klfrank/Dropbox/Post-Grad/Bioinformatics/R-Microbiome-Statistics/Databases/REU-bootcamp/labhuiofrank-16S-util/R/ordination.R")
cca2 <- run_ordination(mb_subsampled, "CCA", formula=~ DO_percent + Temp_C + Sal_ppt + pH)
plot_ordination_with_arrows(cca2$data, cca2$arrows, "CCA1", "CCA2", fill=target) 

```
